### ISSUE TITLE
Make doctest available to plugins

### ISSUE THREAD

**Issue by rsmmr** [zeek/zeek#1477]
We support unit testing through doctests already, however that's currently not directly usable from plugins as initialization order does't quite work out.

**Comment by timwoj** [comment]
I wonder if this would actually work once https://github.com/zeek/zeek/pull/1416 lands.

**Comment by rsmmr** [comment]
That has landed now, do you think it works now, even for dynamic plugins?

**Comment by ckreibich** [comment]
I still have a test plugin sitting around from when I tried this a while back, and found that it's only the Zeek initialization being slightly out of order for making this Just Work. I'll see what I can find.

**Comment by ckreibich** [comment]
I have a few updates here. Unit tests in plugins still do not work out of the box, but it's not hard to get them to work. I made small changes to the Zeek startup logic in a branch: 

https://github.com/zeek/zeek/compare/topic/christian/doctest-in-packages

This just moves the unit test execution down past the plugin initialization. I found that at that point the existing unit tests were no longer deterministic, but enforcing deterministic mode fixed that. It also installs the doctest header.

Here's a Zeek package that work with & without unit test support, and automatically builds with unit tests support when Zeek has it:

https://github.com/ckreibich/testpackage

It just includes the doctest header and has similar configure/cmake goop as Zeek itself for configuring doctest.

```
$ zeek --test
[doctest] doctest version is "2.3.5"
[doctest] run with "--help" for options
error in /home/christian/inst/opt/zeek/share/zeek/base/init-bare.zeek, line 1: string without NUL terminator: "abcdef"
error in /home/christian/inst/opt/zeek/share/zeek/base/init-bare.zeek, line 1: string with embedded NUL: "ab\x00def"
===============================================================================
[doctest] test cases:     58 |     58 passed |      0 failed |      0 skipped
[doctest] assertions:    439 |    439 passed |      0 failed |
[doctest] Status: SUCCESS!
```
In a local clone of the package:
```
$ zkg install .
The following packages will be INSTALLED:
  /home/christian/t1/utest (master)

Proceed? [Y/n] 
Running unit tests for "/home/christian/t1/utest"
Installing "/home/christian/t1/utest".
Installed "/home/christian/t1/utest" (master)
Loaded "/home/christian/t1/utest"
$ zeek -NN Corelight::UnitTest
Corelight::UnitTest - TODO: Insert description (dynamic, version 0.1.0)
```
The package's unit test is now there and runs:
```
$ zeek --test -ltc
[doctest] listing all test case names
===============================================================================
dict construction
...
util json_escape_utf8
plugin doctest
===============================================================================
$ zeek --test
[doctest] doctest version is "2.3.5"
[doctest] run with "--help" for options
error in /home/christian/inst/opt/zeek/share/zeek/base/init-bare.zeek, line 1: string without NUL terminator: "abcdef"
error in /home/christian/inst/opt/zeek/share/zeek/base/init-bare.zeek, line 1: string with embedded NUL: "ab\x00def"
===============================================================================
[doctest] test cases:     59 |     59 passed |      0 failed |      0 skipped
[doctest] assertions:    440 |    440 passed |      0 failed |
[doctest] Status: SUCCESS!
```
But this was all a bit rushed ... not sure about other platforms, for example. I'm leaning toward leaving this out of 4.1, but it seems safe to commit for 4.2.

**Comment by timwoj** [comment]
I think that looks fine to me to merge into 4.1. As long as the tests still run on CI and nothing breaks. It'd be good to add something to one of our test plugins just to make sure it has coverage too.

**Comment by ckreibich** [comment]
Closing, resolved via #1661.
